src/shadowfax.py

```diff
@@ -0,0 +1,405 @@
+import asyncio
+from collections import defaultdict
+import copy
+import uuid
+import hashlib
+import os
+import networkx
+import httpx
+
+from reasoner_pydantic import Message
+
+from src.pathfinder.get_cooccurrence import get_the_curies, get_the_pmids
+
+node_norm_url = os.environ.get("NODENORM_URL", "https://nodenormalization-sri.renci.org/")
+strider_url = os.environ.get("STRIDER_URL", "https://strider.renci.org/")
+num_intermediate_hops = 3
+TOTAL_PUBS = 27840000
+
+async def generate_from_strider(message):
+    try:
+        async with httpx.AsyncClient(timeout=3600) as client:
+            lookup_response = await client.post(
+                url=strider_url + "query",
+                json=message,
+            )
+            lookup_response.raise_for_status()
+            lookup_response = lookup_response.json()
+    except:
+        lookup_response = {}
+    return lookup_response.get("message", {})
+
+
+async def shadowfax(message, guid, logger):
+    qgraph = message["message"]["query_graph"]
+    pinned_node_ids = []
+    for node in qgraph["nodes"].values():
+        if node.get("ids", None) is not None:
+            pinned_node_ids.append(node["ids"][0])
+        else:
+            unpinned_node_category = node.get("categories", ["biolink:NamedThing"])[0]
+    if len(pinned_node_ids) != 2:
+        logger.info(f"{guid}: Pathfinder queries requrie two pinned nodes.")
+        return message, 500
+
+    async with httpx.AsyncClient(timeout=900) as client:
```

Comment:
Repeated below, should it be a function?

Reply:
That's probably a good idea, yeah.